pkg/csr/csr.go

```diff
@@ -0,0 +1,357 @@
+		// hence we should not proceed if a corresponding node already exists
+		node := &core.Node{}
+		err := a.client.Get(context.TODO(), kubeTypes.NamespacedName{Namespace: a.namespace,
+			Name: nodeName}, node)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return false, errors.Wrapf(err, "unable to get node %s", nodeName)
+		} else if err == nil {
+			return false, errors.Wrapf(err, "%s node already exists, cannot validate CSR", nodeName)
+		}
+	} else {
+		if validated, err := a.validateKubeletServingCSR(parsedCSR); !validated {
+			return false, errors.Wrapf(err, "unable to validate kubelet serving CSR %s", a.csr.Name)
+		}
+	}
+	return true, nil
+}
+
+// validateNodeName returns true if the node name passed here matches either the
+// actual host name of the VM'S or the reverse lookup of the instance addresses
+// present in the configMap. If the host name matches, it also validates if the
+// node name complies with the DNS RFC1123 naming convention.
+func (a *Approver) validateNodeName(nodeName string) (bool, error) {
+	configMap := &core.ConfigMap{}
+	err := a.client.Get(context.TODO(), kubeTypes.NamespacedName{Namespace: a.namespace,
+		Name: parser.InstanceConfigMap}, configMap)
+	if err != nil && !k8sapierrors.IsNotFound(err) {
+		return false, errors.Wrapf(err, "could not retrieve Windows instance configMap %s",
+			parser.InstanceConfigMap)
+	}
+
+	nodes := &core.NodeList{}
+	if err := a.client.List(context.TODO(), nodes); err != nil {
+		return false, errors.Wrap(err, "error listing nodes")
+	}
+
+	// Get the list of instances that are expected to be Nodes
+	windowsInstances, err := parser.ParseInstances(configMap.Data, nodes)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to parse hosts from ConfigMap %s", configMap.Name)
+	}
+	// check if the node name matches the lookup of any of the instance addresses
+	hasEntry, err := hasAssociatedInstance(windowsInstances, nodeName)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to map node name to the instances present in the configMap %s",
+			configMap.Name)
+	}
+	if hasEntry {
+		return true, nil
+	}
+	// Create a new signer using the private key secret
+	instanceSigner, err := signer.Create(kubeTypes.NamespacedName{Namespace: a.namespace,
+		Name: secrets.PrivateKeySecret}, a.client)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to create signer from private key secret")
+	}
+	// check if the node name matches any of the instance's host names
+	hasEntry, err = hasAssociatedHost(nodeName, windowsInstances, instanceSigner)
+	if err != nil {
+		return false, errors.Wrapf(err, "unable to map node name to the host names of instances in configMap %s",
+			configMap.Name)
+	}
+	if !hasEntry {
+		return false, nil
+	}
+	// validate node name for DNS RFC1123 naming conventions
+	// ref: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names
```

Comment:
Are you sure nodes names need to follow this standard or the more strict DNS label standards? https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names